Add styling to marketplace to identify sponsored products

Dan-Q opened the pull request:
### Changes proposed in this Pull Request:

Closes [WCCOM#19612](https://github.com/Automattic/woocommerce.com/issues/19612). Where a product on the Marketplace "Discover" page has the `label: 'promoted'` from the WooCom API, adds the word "Sponsored" to the product card and shows a colored spot alongside/overlapping it. If it also has a `primary_color: ...` set, adds a "stripe" of that primary color to the top of the product card.

### How to test the changes in this Pull Request:

1. Check out this branch to your local environment and build the assets, e.g. `cd plugins/woocommerce-admin; pnpm watch:build`; visit the site in your local server e.g. `cd plugins/woocommerce; pnpm -- wp-env start`
2. If you haven't done it in a while, consider clearing your `wc_addons_featured` transient e.g. `cd plugins/woocommerce; pnpm -- wp-env run cli wp transient delete wc_addons_featured`; this transient stores featured product data
3. [Visit WooCommerce > Extensions > Discover](http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions)
4. If any shown products are marked "promoted" in WooCom, they should appear here in accordance with [the Figma design](https://www.figma.com/file/PajqJjAVxHoQrSh0kXBpiJ/Ads?type=design&node-id=29-4028&mode=design&t=ErLpXi2Gnap5kXhI-0).
5. If you'd like to test around or if the data you're seeing shows no sponsored products, you can use your React debugger to modify the `product` associated with each `<ProductCard>`
6. Please test around

#### Screenshot

![Markup on 2024-03-18 at 17:34:06](https://github.com/woocommerce/woocommerce/assets/53293/6bbba577-5e08-4182-98d6-259a4453c313)

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Made sponsored product listings in the Extensions marketplace easier to identify.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Dan-Q committed:
Pass 'Sponsored' detail down to product cards from API

Dan-Q committed:
Style a "Sponsored" label

Dan-Q committed:
Stripe on top of sponsored products

Dan-Q committed:
Add 'blob' next to sponsored listings

Dan-Q committed:
Fix indentation bug

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

poligilad-auto commented:
Hey @Dan-Q, thanks for the ping.

I didn't get a chance to test the branch, but from the screenshot I can see a few things:

- The spacing between the product title and vendor name is [too big](https://prnt.sc/aQmi0SfAaMA8) vs. [how it is now](https://prnt.sc/AuFStRRnZQCD).
- Can you check that the top stripe is 5px? It [currently](https://prnt.sc/OIvMeVhppXI8) looks a bit higher.
- This is probably not related and maybe a DEV environment bug, but I see the ["Didn't find the theme you like? ..."](https://prnt.sc/i8fgpSf_UdeF) sentence repeated twice and it should only be shown at the end of all the Themes in `Extensions > Themes` and not shown at all in `Discover`.

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -33,13 +36,33 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+			return {};
+		}
+		return {
+			background: `linear-gradient(${product.primary_color} 0, ${product.primary_color} ${SPONSORED_PRODUCT_STRIPE_SIZE}, white ${SPONSORED_PRODUCT_STRIPE_SIZE}, white)`
```
This line gives a lint.

```suggestion
			background: `linear-gradient(${ product.primary_color } 0, ${ product.primary_color } ${ SPONSORED_PRODUCT_STRIPE_SIZE }, white ${ SPONSORED_PRODUCT_STRIPE_SIZE }, white)`,

```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -33,13 +36,33 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+	 * In an ideal world this could be set in a data- attribute and we'd use CSS calc() and attr() to get it, but
+	 * attr() doesn't have very good support yet, so we need to apply some inline CSS to stripe sponsored results.
+	 */
+	function inlineCss(): Object {
```
```suggestion
	function inlineCss(): object {
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -104,11 +127,12 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
 
 	return (
-		<Card className={ classNames } aria-hidden={ isLoading }>
+		<Card className={ classNames } aria-hidden={ isLoading } style={ inlineCss() }>
```
```suggestion
		<Card
			className={ classNames }
			aria-hidden={ isLoading }
			style={ inlineCss() }
		>
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -158,14 +182,28 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+									<span className="woocommerce-marketplace__product-card__vendor" />
 								</p>
 							) }
+							{ ! isLoading &&
```
```suggestion
							{ ! isLoading && (
```

corsonr replied:
If you add this one you need to close it on line 210

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -158,14 +182,28 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
 							) }
+							{ ! isLoading &&
+								<p className="woocommerce-marketplace__product-card__vendor-details">
+									{ productVendor &&
```
```suggestion
									{ productVendor && (
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -158,14 +182,28 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+											<span>{ __( 'By ', 'woocommerce' ) }</span>
+											{ productVendor }
+										</span>
+									}
```
```suggestion
									) }
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -158,14 +182,28 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+								<p className="woocommerce-marketplace__product-card__vendor-details">
+									{ productVendor &&
+										<span className="woocommerce-marketplace__product-card__vendor">
+											<span>{ __( 'By ', 'woocommerce' ) }</span>
```
```suggestion
											<span>
												{ __( 'By ', 'woocommerce' ) }
											</span>
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -158,14 +182,28 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+											{ productVendor }
+										</span>
+									}
+									{ productVendor && isSponsored() &&
```
```suggestion
									{ productVendor && isSponsored() && (
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -158,14 +182,28 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+									}
+									{ productVendor && isSponsored() &&
+										<span aria-hidden="true" className="woocommerce-marketplace__product-card__vendor-details__separator">·</span>
+									}
```
```suggestion
									) }
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -158,14 +182,28 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+										</span>
+									}
+									{ productVendor && isSponsored() &&
+										<span aria-hidden="true" className="woocommerce-marketplace__product-card__vendor-details__separator">·</span>
```
```suggestion
										<span
											aria-hidden="true"
											className="woocommerce-marketplace__product-card__vendor-details__separator"
										>
											·
										</span>
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -158,14 +182,28 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+									{ productVendor && isSponsored() &&
+										<span aria-hidden="true" className="woocommerce-marketplace__product-card__vendor-details__separator">·</span>
+									}
+									{ isSponsored() &&
```
```suggestion
									{ isSponsored() && (
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -158,14 +182,28 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
+										<span className="woocommerce-marketplace__product-card__sponsored-label">
+											{ __( 'Sponsored', 'woocommerce' ) }
+										</span>
+									}
```
```suggestion
									) }
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx:
```diff
@@ -33,13 +36,33 @@ function ProductCard( props: ProductCardProps ): JSX.Element {
 		vendorUrl: '',
 		icon: '',
+		label: null,
+		primary_color: null,
```
Any reason to use `primaryColor` instead of `primary_color` for consistency?

Dan-Q replied:
The object gets `primary_color` as standard (that's what comes from the API, for some reason). I initially remapped it with `primaryColor: product.primary_color` but I reasoned that this was actually worse because we were now using _two_ different names for the same thing!

Consistency within the variable seemed to be more-important than consistency between different variables.

corsonr reviewed (commented):
I'm getting a couple of lints, so I left a couple of suggestions.

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -269,6 +284,30 @@
+			content: ' ';
+			height: 46px;
+			width: 46px;
+			background: #09B58540;
```
Maybe this should be added to `../../stylesheets/_variables.scss`?

Dan-Q replied:
Done in 920444cecd9d0f7d288caff374fa7a443b3da50a. Thanks!

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -269,6 +284,30 @@
+			content: ' ';
+			height: 30px;
+			width: 30px;
+			background: #09B585;
```
Maybe this should be added to `../../stylesheets/_variables.scss`?

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -269,6 +284,30 @@
 		}
+
+		&.is-sponsored::before {
+			content: ' ';
+			height: 46px;
+			width: 46px;
+			background: #09B58540;
+			display: block;
+			border-radius: 50%;
+			position: absolute;
+			right: -23px;
+			top: 20px;
```
Let's sort
```suggestion
			background: #09B58540;
			border-radius: 50%;
			content: ' ';
			display: block;
			height: 46px;
			position: absolute;
			right: -23px;
			top: 20px;
			width: 46px;
```

corsonr commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -269,6 +284,30 @@
+		}
+
+		&.is-sponsored::after {
+			content: ' ';
+			height: 30px;
+			width: 30px;
+			background: #09B585;
+			display: block;
+			border-radius: 50%;
+			position: absolute;
+			right: -15px;
+			top: 28px;
```
Let's sort
```suggestion
			background: #09B585;
			border-radius: 50%;
			content: ' ';
			display: block;
			height: 30px;
			position: absolute;
			right: -15px;
			top: 28px;
			width: 30px;
```

Dan-Q committed:
Update plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx

Co-authored-by: Remi Corson <remicorson@gmail.com>

Dan-Q committed:
Update plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx

Co-authored-by: Remi Corson <remicorson@gmail.com>

Dan-Q committed:
Update plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.tsx

Co-authored-by: Remi Corson <remicorson@gmail.com>

Dan-Q committed:
Apply suggestions from code review

Co-authored-by: Remi Corson <remicorson@gmail.com>

Dan-Q committed:
Add missing close parenthesis following linter changes

Dan-Q committed:
Make linter happier about hex colors (still needs variablising)

Dan-Q committed:
Fix margins around vendor name

Dan-Q committed:
Fixed height of product stripe

Dan-Q commented:
> I didn't get a chance to test the branch, but from the screenshot I can see a few things:

It's funny: you stare at something long enough, you stop seeing the problems!

>     * The spacing between the product title and vendor name is [too big](https://prnt.sc/aQmi0SfAaMA8) vs. [how it is now](https://prnt.sc/AuFStRRnZQCD).

Fixed in 3b61fcc206f01d2f60f7e8671e3c2b603cc6011e.

>     * Can you check that the top stripe is 5px? It [currently](https://prnt.sc/OIvMeVhppXI8) looks a bit higher.

Fixed in 3121d3077983cce5c2b3cffef1d0b099f4d57626.

>     * This is probably not related and maybe a DEV environment bug, but I see the ["Didn't find the theme you like? ..."](https://prnt.sc/i8fgpSf_UdeF) sentence repeated twice and it should only be shown at the end of all the Themes in `Extensions > Themes` and not shown at all in `Discover`.

Huh; I see that too, but I don't think it's new to this PR... investigating.

Dan-Q committed:
Linting

Dan-Q commented:
> Huh; I see that too, but I don't think it's new to this PR... investigating.

[This is a regression introduced a week ago](https://github.com/woocommerce/woocommerce/pull/45468); @nefeline is handling.

Dan-Q committed:
Moved color to variable and calculated alpha using Sass functions

Dan-Q commented:
Revised version screenshot:

![Markup on 2024-03-19 at 11:35:55](https://github.com/woocommerce/woocommerce/assets/53293/5fefe589-90da-4205-8e89-18a9e197c3cb)

Dan-Q commented:
> One last thing, I'm not getting the sponsored card to show even when manipulating the props via the label prop set to `promoted` and by setting a background color.

Here's me doing so - see if doing this works for you:

https://github.com/woocommerce/woocommerce/assets/53293/4a95a777-8807-43c4-b079-fc88bdebe4a3

corsonr reviewed (approved):
Thanks for all the changes, that works beautifully!

<img width="1082" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1649788/2d0d964a-4d49-416a-b013-9cd36bee1613">

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -269,6 +284,30 @@
 				line-height: 1.5;
 			}
 		}
+
+		&.is-sponsored::before {
+			background: rgba($celadon-30, 0.25);
+			border-radius: 50%;
+			content: ' ';
+			display: block;
+			height: 46px;
+			position: absolute;
+			right: -23px;
+			top: 20px;
+			width: 46px;
+		}
+
+		&.is-sponsored::after {
+			background: $celadon-30;
+			border-radius: 50%;
+			content: ' ';
+			display: block;
+			height: 30px;
+			position: absolute;
+			right: -15px;
+			top: 28px;
+			width: 30px;
+		}
```
```suggestion
```

Turns out this is a designers' mark, not design. Whoops!

Dan-Q committed:
Update plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/stylesheets/_variables.scss:
```diff
@@ -33,6 +33,7 @@ $wp-gray-0: $gray-0;
 $wp-gray-50: $gray-50;
 $wp-gray-60: $gray-60;
 $skeleton-loader-color: #f0f0f0;
+$celadon-30: #09b585;
```
```suggestion
```

No longer needed.

Dan-Q committed:
Update plugins/woocommerce-admin/client/marketplace/stylesheets/_variables.scss

Dan-Q commented:
Turns out the green blobs aren't part of the design, whoops! Fixed in 0b11add79477224f671e0e4c015092f757e799c8.

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card.scss:
```diff
@@ -181,6 +192,10 @@
 			text-decoration: none;
 		}
 
+		&__sponsored-label {
+			color: $gray-700;
+		}
```
This could maybe have been `woocommerce-marketplace__product-card__label`, as we may show labels with a different text at some point. Not a big deal I guess.

andfinally reviewed (approved):
Looks great, thanks Dan.
